core/lib/types/src/protocol_version.rs

```diff
@@ -408,6 +434,91 @@ impl TryFrom<Log> for ProtocolUpgrade {
     }
 }
 
+impl TryFrom<Log> for GovernanceOperation {
+    type Error = crate::ethabi::Error;
+
+    fn try_from(event: Log) -> Result<Self, Self::Error> {
+        let call_param_type = ParamType::Tuple(vec![
+            ParamType::Address,
+            ParamType::Uint(256),
+            ParamType::Bytes,
+        ]);
+
+        let operation_param_type = ParamType::Tuple(vec![
+            ParamType::Array(Box::new(call_param_type)),
+            ParamType::FixedBytes(32),
+            ParamType::FixedBytes(32),
+        ]);
+        let mut decoded = decode(&[ParamType::Uint(256), operation_param_type], &event.data.0)?;
+        decoded = decoded.remove(1).into_tuple().unwrap();
+
+        let eth_hash = event
+            .transaction_hash
+            .expect("Event transaction hash is missing");
+        let eth_block = event
+            .block_number
+            .expect("Event block number is missing")
+            .as_u64();
+
+        let calls = decoded.remove(0).into_array().unwrap();
+        let predecessor = H256::from_slice(&decoded.remove(0).into_fixed_bytes().unwrap());
+        let salt = H256::from_slice(&decoded.remove(0).into_fixed_bytes().unwrap());
+
+        let calls = calls
+            .into_iter()
+            .map(|call| {
+                let mut decoded = call.into_tuple().unwrap();
+
+                Call {
+                    target: decoded.remove(0).into_address().unwrap(),
+                    value: decoded.remove(0).into_uint().unwrap(),
+                    data: decoded.remove(0).into_bytes().unwrap(),
+                    eth_hash,
+                    eth_block,
+                }
+            })
+            .collect();
+
+        Ok(Self {
+            calls,
+            predecessor,
+            salt,
+        })
+    }
+}
+
+impl TryFrom<Call> for ProtocolUpgrade {
+    type Error = crate::ethabi::Error;
+
+    fn try_from(call: Call) -> Result<Self, Self::Error> {
+        // Reuses `ProtocolUpgrade::try_from`.
+        // `ProtocolUpgrade::try_from` only uses 3 log fields: `data`, `block_number`, `transaction_hash`. Others can be filled with dummy values.
+        // We build data as `call.data` without first 4 bytes which are for selector
+        // and append it with `bytes32(0)` for compatibility with old event data.
+        let data = call
+            .data
+            .into_iter()
+            .skip(4)
+            .chain(encode(&[Token::FixedBytes(H256::zero().0.to_vec())]))
+            .collect::<Vec<u8>>()
+            .into();
+        let log = Log {
+            address: Default::default(),
+            topics: Default::default(),
+            data,
+            block_hash: Default::default(),
```

Comment:
Nit: 
..Default::default()

Reply:
It requires `Default` to be implemented for Log, which is not the case